src/cli/services/linter.ts

```diff
@@ -0,0 +1,128 @@
+
+export async function loadRulesets(cwd: string, rulesetFiles: string | string[] | null): Promise<IRuleset> {
+  if (rulesetFiles === null || rulesetFiles.length === 0) {
+    return {
+      functions: {},
+      rules: {},
+    };
+  }
+
+  return readRuleset(
+    (Array.isArray(rulesetFiles) ? rulesetFiles : [rulesetFiles]).map(
+      file => (isAbsolute(file) ? file : resolve(cwd, file)),
+    ),
+  );
+}
+
+export async function lint(name: string, flags: ILintConfig, rulesetFile: Optional<string | string[]>) {
+  if (flags.verbose) {
+    console.info(`Linting ${name}`);
+  }
+
+  let targetUri = name;
+  if (!/^https?:\/\//.test(name) && !isAbsolute(name)) {
+    // we always want the absolute path to the target file
+    targetUri = resolve(process.cwd(), name);
+  }
+
+  const spec: IParserResult = parseWithPointers(await readParsable(targetUri, { encoding: flags.encoding }), {
+    ignoreDuplicateKeys: false,
+    mergeKeys: true,
+  });
+
+  const rulesetFiles = rulesetFile || (await getDefaultRulesetFile(process.cwd()));
+  const ruleset = await (rulesetFiles ? loadRulesets(process.cwd(), rulesetFiles) : readRuleset('spectral:oas'));
+
+  const spectral = new Spectral({ resolver: httpAndFileResolver });
+
+  if (flags.verbose) {
+    if (ruleset) {
+      console.info(`Found ${Object.keys(ruleset.rules).length} rules`);
+    } else {
+      console.info('No rules loaded, attempting to detect document type');
+    }
+  }
+
+  spectral.registerFormat('oas2', document => {
+    if (isOpenApiv2(document)) {
+      if (!flags.quiet) {
+        console.log('OpenAPI 2.0 (Swagger) detected');
+      }
+
+      return true;
+    }
+    return false;
+  });
+
+  spectral.registerFormat('oas3', document => {
+    if (isOpenApiv3(document)) {
+      if (!flags.quiet) {
+        console.log('OpenAPI 3.x detected');
+      }
+
+      return true;
+    }
+    return false;
+  });
+
+  spectral.setRuleset(ruleset);
+
+  if (flags.skipRule) {
+    spectral.setRules(skipRules(ruleset.rules, flags));
+  }
+
+  const parsedResult: IParsedResult = {
+    source: targetUri,
+    parsed: spec,
+    getLocationForJsonPath,
+  };
+
+  return spectral.run(parsedResult, {
+    resolve: {
+      documentUri: targetUri,
+    },
+  });
+}
+
+const skipRules = (rules: RuleCollection, flags: ILintConfig): RuleCollection => {
+  const skippedRules: string[] = [];
+  const invalidRules: string[] = [];
+
+  if (flags.skipRule !== undefined) {
+    for (const rule of Array.isArray(flags.skipRule) ? flags.skipRule : [flags.skipRule]) {
+      if (rule in rules) {
+        delete rules[rule];
+        skippedRules.push(rule);
+      } else {
+        invalidRules.push(rule);
+      }
+    }
+  }
+
+  if (invalidRules.length !== 0) {
+    if (!flags.quiet) {
+      console.warn(`ignoring invalid ${invalidRules.length > 1 ? 'rules' : 'rule'} "${invalidRules.join(', ')}"`);
+    }
+  }
+
+  if (skippedRules.length !== 0 && flags.verbose) {
+    if (flags.verbose) {
```

Comment:
Interesting stuff here — in theory this could be handled by a logger that will decide whether output something or not based on its configuration; so you can always output stuff and not having to use ifs everywhere. Could be a good next step

Reply:
> y a logger that will decide whether output something or not based on its configuration; 

Yeah, I have a basic implementation of Logger already. Going to ship it in 4.2